### ISSUE TITLE
V2.4.6 crashes with UndefinedMethodException using an imported macro

### ISSUE THREAD

**Issue by abzolv** [twigphp/Twig#2648]
Symfony 3.4.5, local dev machine, Twig v2.4.6

At the top of the view file:
{% from 'macros/formmacros.html.twig' import form_button_row, form_cancel_button %}

Error:
UndefinedMethodException
Attempted to call an undefined method named "macro_form_button_row" of class "__TwigTemplate_883c90ebfb6b93c87507a2c904fa55cbb5eb1b22de3438727a67f0eda99bc3e9".

This works unchanged in V2.4.4 and prior.

**Comment by jmsche** [comment]
I can confirm the bug using the same version of Twig.

Reverting this commit seems to fix the bug: https://github.com/twigphp/Twig/commit/6ab5fe9b8fd52f810fd902652659605b5fcf39a9

**Comment by stof** [comment]
I tried to reproduce the issue like that https://twigfiddle.com/hkck2n but it works fine with Twig 2.4.6.

Can you give more details about the reproducing case ?

**Comment by abzolv** [comment]
Does TwigFiddle also compile the code before execution in exactly the same way as Symfony does?

I can't reproduce the error on TwigFiddle, even though I make the code an exact match of what I use in Symfony.

As you'll notice from the original post, the error occurs in the compiled code in the /var/cache/dev folder of Symfony.

**Comment by stof** [comment]
Well, the template compilation is something done by Twig itself. Symfony only configures its own cache dir as the place to store the Twig cache.

If this only happens in a Symfony project and not in Twigfiddle, it might be related to extra Twig extensions registered in the project compared to TwigFiddle.

Can you share a reproducing case, even if it involves using a Symfony project ?

**Comment by abzolv** [comment]
Not sure if it will help, but the line in the compiled file where the crash happens reads:

```
echo $context["__internal_7f693234e20aef99a0a23c623562dc1dd3f22f11bd95d2d0e71db1af48b3ecc5"]->macro_form_button_row(twig_get_attribute($this->env, $this->source, (isset($context["form"])
```

**Comment by abzolv** [comment]
Not sure it it will help, but the line in the compiled file where the crash happens reads:

```
echo $context["__internal_7f693234e20aef99a0a23c623562dc1dd3f22f11bd95d2d0e71db1af48b3ecc5"]->macro_form_button_row(array(0 => twig_get_attribute($this->env, $this->source, (isset($context["bootstrap_service"]) || array_key_exists("bootstrap_service", $context) ? $context["bootstrap_service"] : (function () { throw new Twig_Error_Runtime('Variable "bootstrap_service" does not exist.', 81, $this->source); })()), "formButton", array(0 => array("title" => $this->extensions['Symfony\Bridge\Twig\Extension\TranslationExtension']->trans("Save"), "icon" => "save", "type" => "primary")), "method"), 1 => twig_get_attribute($this->env, $this->source, (isset($context["bootstrap_service"]) || array_key_exists("bootstrap_service", $context) ? $context["bootstrap_service"] : (function () { throw new Twig_Error_Runtime('Variable "bootstrap_service" does not exist.', 81, $this->source); })()), "linkButton", array(0 => array("title" => $this->extensions['Symfony\Bridge\Twig\Extension\TranslationExtension']->trans("action.cancel"), "url" => $this->extensions['Symfony\Bridge\Twig\Extension\RoutingExtension']->getPath("queues_list"), "icon" => "times")), "method")));
```

**Comment by abzolv** [comment]
As I mentioned earlier, it's the delta between 2.4.4 and 2.4.6 that's causing the problem.

**Comment by stof** [comment]
That's still not giving us an input reproducing the issue, meaning we cannot debug it.

**Comment by abzolv** [comment]
Here's a Symphony project that reproduces the problem:

https://github.com/abzolv/twigcrashdemo

The crash happens when the leaf twig view (that imports the macro) extends another twig view, which imports another macro. In the demo's case, the second import is in base.html.twig.

**Comment by liverbool** [comment]
Me too!

```
symfony/symfonyv 3.4.6
twig/twig v2.4.6
```

**Comment by viovet-dan** [comment]
Me too!

```
symfony 3.3.16
twig 2.4.6
```

Going back to 2.4.4 fixes the issue.

**Comment by stof** [comment]
> The crash happens when the leaf twig view (that imports the macro) extends another twig view, which imports another macro

thanks. That was the missing bit.

@fabpot should we use a hash of the imported template to create the variable for the macro import instead (this way, the name would be the same only if the child template imports the same macros, which would not break due to overriding the variable)

**Comment by fabpot** [comment]
fixed by #2664
